src/lib/reasoners/bitv.ml

```diff
@@ -289,40 +315,557 @@ module Shostak(X : ALIEN) = struct
+                t, nctx
+            | None ->
+              let t1 = { bv = I_Other r; sz = 1} in
+              if ncnt > size then t1, nctx
+              else
+                let t2, nctx =
+                  aux ncnt 0 (mk_nthbv_eq bv_t cnt rv_t :: nctx)
+                in
+                let t = mk_i_comp t2 t1 (t2.sz + 1) in
+                t, nctx
+          end
+      in
+      aux 0 0 ctx
+
+    let bvor_ctes ~is_and b1 sz1 b2 sz2 =
+      let bres = if is_and then b1 && b2 else b1 || b2 in
+      if sz1 > sz2 then
+        {bv = Cte bres; sz = sz2}, [{ bv = Cte b1; sz = sz1 - sz2 }], []
+      else if sz1 < sz2 then
+        {bv = Cte bres; sz = sz1}, [], [{ bv = Cte b1; sz = sz2 - sz1 }]
+      else {bv = Cte bres; sz = sz1}, [], []
+
+    let bvor_cte_other ~is_and b sz1 o sz2 =
+      let bres = if is_and then not b else b in
+      if sz1 > sz2 then
+        let l1 = [{ bv = Cte b; sz = sz1 - sz2 }] in
+        if bres then { bv = Cte b; sz = sz2 }, l1, []
+        else { bv = Other o; sz = sz2 }, l1, []
+      else if sz1 < sz2 then
+        let l2 = [{ bv = Ext (o, sz2, 0, sz2 - sz1 - 1); sz = sz2 - sz1 }] in
+        if bres then { bv = Cte b; sz = sz1 }, [], l2
+        else { bv = Ext (o, sz2, sz2 - sz1, sz2 - 1); sz = sz1 }, [], l2
+      else
+      if bres then { bv = Cte b; sz = sz1 }, [], []
+      else { bv = Other o; sz = sz1 }, [], []
+
+    let bvor_cte_ext ~is_and b sz1 o sz2' sz2 lb ub =
+      let bres = if is_and then not b else b in
+      if sz1 > sz2 then
+        let l1 = [{ bv = Cte b; sz = sz1 - sz2 }] in
+        if bres
+        then { bv = Cte b; sz = sz2 }, l1, []
+        else { bv = Ext (o, sz2', lb, ub); sz = sz2 }, l1, []
+      else if sz1 < sz2 then
+        let l2 =
+          [{ bv = Ext (o, sz2', lb, lb + sz2 - sz1 - 1); sz = sz2 - sz1 }]
+        in
+        if bres
+        then { bv = Cte b; sz = sz1 }, [], l2
+        else { bv = Ext (o, sz2', lb + sz2 - sz1, ub); sz = sz1 }, [], l2
+      else
+      if bres
+      then { bv = Cte b; sz = sz1 }, [], []
+      else { bv = Ext (o, sz2', lb, ub); sz = sz1 }, [], []
+
+    (** [mk_bvget_or_eq t cnt t1 cnt1 t2 cnt2]:
+        t[cnt] = if t1[cnt1] = 0 then t2[cnt2] else 1 *)
+    let mk_bvget_or_eq nt cnt t1 cnt1 t2 cnt2 =
+      E.mk_eq ~iff:false (E.mk_bitv_extract cnt cnt nt 1) (
+        E.mk_ite
+          (E.mk_eq ~iff:false (E.mk_bitv_extract cnt1 cnt1 t1 1) E.bvzero)
+          (E.mk_bitv_extract cnt2 cnt2 t2 1)
+          E.bvone)
+
+    (** [mk_bvget_and_eq t cnt t1 cnt1 t2 cnt2]:
+        t[cnt] = if t1[cnt1] = 1 then t2[cnt2] else 0 *)
+    let mk_bvget_and_eq nt cnt t1 cnt1 t2 cnt2 =
+      E.mk_eq ~iff:false (E.mk_bitv_extract cnt cnt nt 1) (
+        E.mk_ite
+          (E.mk_eq ~iff:false (E.mk_bitv_extract cnt1 cnt1 t1 1) E.bvone)
+          (E.mk_bitv_extract cnt2 cnt2 t2 1)
+          E.bvzero
+      )
+
+    (** turn into a dispatcher function  *)
+    let mk_o_eqs =
```

Comment:
Can you add a comment clarifying what this is doing?

Reply:
Done